### PR TITLE
feat: Add payment-processor for any to erc20 payment network 

### PR DESCRIPTION
**Pull request by vrolland**

### CODE REVIEW

File: packages/payment-processor/src/payment/any-to-erc20-proxy.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    to: proxyAddress,
+    value: 0,
+    ...overrides,
+  });
+
+  return tx;
+}
+
+/**
+ * Encodes the call to pay a request through the ERC20 fee proxy contract, can be used with a Multisig contract.
```
Comment (yomarion) [inline packages/payment-processor/src/payment/any-to-erc20-proxy.ts:63]:
```suggestion
 * Encodes the call to pay a request with an ERC20 currency that is different from the request currency (eg. fiat). The payment is made by the ERC20 fee proxy contract.
```

File: packages/payment-processor/src/payment/any-to-erc20-proxy.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import {
+  getAmountToPay,
+  getProvider,
+  getRequestPaymentValues,
+  getSigner,
+  validateConversionFeeProxyRequest,
+} from './utils';
+
+/**
+ * Processes a transaction to pay an ERC20 Request with fees.
```
Comment (yomarion) [inline packages/payment-processor/src/payment/any-to-erc20-proxy.ts:20]:
```suggestion
 * Processes a transaction to pay a request with an ERC20 currency that is different from the request currency (eg. fiat). The payment is made by the ERC20 fee proxy contract.
```

**Review by yomarion — COMMENTED** [review]
Sorry, partial review, I need to continue later.

File: packages/payment-processor/src/payment/conversion-erc20.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+import { _getErc20FeeProxyPaymentUrl } from './erc20-fee-proxy';
+import { _getErc20ProxyPaymentUrl } from './erc20-proxy';
+
+import { ITransactionOverrides } from './transaction-overrides';
+import { getProvider, getSigner } from './utils';
+import { checkErc20Allowance, encodeApproveAnyErc20 } from './erc20';
+
+/**
+ * Processes the approval transaction of a given payment ERC20 to be spent by the swap router,
```
Comment (yomarion) [inline packages/payment-processor/src/payment/conversion-erc20.ts:16]: Please review your comments to match the methods (look for references to router or fee proxy)

File: packages/payment-processor/src/payment/conversion-erc20.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      request,
+      paymentTokenAddress,
+      signerOrProvider,
+      overrides,
+    );
+  }
+}
+
+/**
+ * Processes the approval transaction of the payment ERC20 to be spent by the swap router.
```
Comment (yomarion) [inline packages/payment-processor/src/payment/conversion-erc20.ts:56]: Please review your comments to match the methods (look for references to router or fee proxy)

File: packages/payment-processor/src/payment/utils.ts [hunk 5]
```diff
@@ -161,6 +184,44 @@ export function validateErc20FeeProxyRequest(
...
+  const { feeAddress, feeAmount, tokensAccepted } = getRequestPaymentValues(request);
+  const amountToPay = getAmountToPay(request, amount);
+  const feeToPay = bigNumberify(feeAmountOverride || feeAmount || 0);
+
+  if (!!feeAmount !== !!feeAddress) {
+    throw new Error('Both fee address and fee amount have to be declared, or both left empty');
+  }
+  if (amountToPay.isZero() && feeToPay.isZero()) {
+    throw new Error('Request payment amount and fee are 0');
+  }
```
Comment (yomarion) [inline packages/payment-processor/src/payment/utils.ts:209]: Small factorization here I think with something like `validateFeesAndAmount(request)` used for fee-proxy and any-to-erc20 ?

File: packages/payment-processor/test/payment/any-to-erc20-proxy.test.ts [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        erc20ContractAddress,
+        provider,
+      ).balanceOf(wallet.address);
+
+      // Check each balance
+      expect(bigNumberify(balanceEthBefore).sub(balanceEthAfter).toNumber()).toBeGreaterThan(0);
+      expect(
+        bigNumberify(balanceTokenBefore)
+          .sub(bigNumberify(balanceTokenAfter))
+          .eq(bigNumberify('1211881188118811880')),
```
Comment (yomarion) [inline packages/payment-processor/test/payment/any-to-erc20-proxy.test.ts:192]: Please comment how to verify that this is the number we should expect.

**Review by yomarion — CHANGES_REQUESTED** [review]
Nice changes, we can soon integrate!

File: packages/payment-processor/src/payment/any-to-erc20-proxy.ts [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  const {
+    paymentReference,
+    paymentAddress,
+    feeAddress,
+    feeAmount,
+    maxRateTimespan,
+  } = getRequestPaymentValues(request);
+
+  const chainlinkDecimal = 8;
+  const decimalPadding = chainlinkDecimal - getDecimalsForCurrency(request.currencyInfo);
```
Comment (benjlevesque) [inline packages/payment-processor/src/payment/any-to-erc20-proxy.ts:104]: isn't it possible that this is less than 0?

**Review by benjlevesque — APPROVED** [review]
the intended use of this package is to be payment network agnostic. we also export specific payment network in case an app has a specific behaviour.

Did you forget to add this payment network to the default `payRequest` method, or is it an in-between step?

**Comment by yomarion** [comment]
> the intended use of this package is to be payment network agnostic. we also export specific payment network in case an app has a specific behaviour.
> 
> Did you forget to add this payment network to the default `payRequest` method, or is it an in-between step?

I think you are right @benjlevesque , `payRequest` is a relevant standard interface for payment. We forgot because the reference was swapToPay, which is very different in a way. Can you have a look at these changes now?

File: packages/payment-processor/src/payment/any-to-erc20-proxy.ts (hunk 7 above)
Comment (yomarion) [reply packages/payment-processor/src/payment/any-to-erc20-proxy.ts:119]: It is, I have just tested. I will fix it and keep the test for later. Awesome finding.

**Review by benjlevesque — APPROVED** [review]

**Review by yomarion — APPROVED** [review]

**Pull request merged**